pkg/storage/tsdb/multilevel_chunk_cache.go

```diff
@@ -0,0 +1,149 @@
+package tsdb
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/cache"
+	"github.com/thanos-io/thanos/pkg/cacheutil"
+)
+
+type multiLevelChunkCache struct {
+	name   string
+	caches []cache.Cache
+
+	backfillProcessor    *cacheutil.AsyncOperationProcessor
+	fetchLatency         *prometheus.HistogramVec
+	backFillLatency      *prometheus.HistogramVec
+	storeDroppedItems    prometheus.Counter
+	backfillDroppedItems prometheus.Counter
+	maxBackfillItems     int
+	backfillTTL          time.Duration
+}
+
+type MultiLevelChunkCacheConfig struct {
+	MaxAsyncConcurrency int `yaml:"max_async_concurrency"`
+	MaxAsyncBufferSize  int `yaml:"max_async_buffer_size"`
+	MaxBackfillItems    int `yaml:"max_backfill_items"`
+
+	BackFillTTL time.Duration `yaml:"-"`
+}
+
+func (cfg *MultiLevelChunkCacheConfig) Validate() error {
+	if cfg.MaxAsyncBufferSize <= 0 {
+		return errInvalidMaxAsyncBufferSize
+	}
+	if cfg.MaxAsyncConcurrency <= 0 {
+		return errInvalidMaxAsyncConcurrency
+	}
+	if cfg.MaxBackfillItems <= 0 {
+		return errInvalidMaxBackfillItems
+	}
+	return nil
+}
+
+func (cfg *MultiLevelChunkCacheConfig) RegisterFlagsWithPrefix(f *flag.FlagSet, prefix string) {
+	f.IntVar(&cfg.MaxAsyncConcurrency, prefix+"max-async-concurrency", 50, "The maximum number of concurrent asynchronous operations can occur when backfilling cache items.")
+	f.IntVar(&cfg.MaxAsyncBufferSize, prefix+"max-async-buffer-size", 10000, "The maximum number of enqueued asynchronous operations allowed when backfilling cache items.")
+	f.IntVar(&cfg.MaxBackfillItems, prefix+"max-backfill-items", 10000, "The maximum number of items to backfill per asynchronous operation.")
+}
+
+func newMultiLevelChunkCache(name string, cfg MultiLevelChunkCacheConfig, reg prometheus.Registerer, c ...cache.Cache) cache.Cache {
+	if len(c) == 1 {
+		return c[0]
+	}
+
+	return &multiLevelChunkCache{
+		name:              name,
+		caches:            c,
+		backfillProcessor: cacheutil.NewAsyncOperationProcessor(cfg.MaxAsyncBufferSize, cfg.MaxAsyncConcurrency),
+		fetchLatency: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "cortex_store_multilevel_chunk_cache_fetch_duration_seconds",
```

Comment:
Nit. Can we call it `chunks_cache` instead of `chunk_cache`? THen the naming is consistent with the config file.